R/trackhub.R

```diff
@@ -0,0 +1,633 @@
+        if (!isFieldEmpty(genomesFileValue)) {
+            genomesFilePath <- combineURI(x, genomesFileValue)
+            genomesList <- getGenomesContent(genomesFilePath)
+        }
+        else message("hub.txt: 'genomesFile' does not contain valid reference to genomes file")
+    }
+}
+
+setGenomesKey <- function(x, key, value) {
+    genomesList <- getGenomesContentList(trackhub(x))
+    position <- which(sapply(genomesList, function(y) genome(x) %in% y))
+    genomesList[[position]][[key]] <- value
+    hubContent <- getHubContent(hubFile(trackhub(x)))
+    genomesFilePath <- combineURI(trackhub(x), hubContent["genomesFile"])
+    cat("", file = genomesFilePath)
+    sapply(genomesList, function(x) {
+        setGenomeContentList(x, genomesFilePath)
+    })
+    return(invisible(NULL))
+}
+
+getGenomesKey <- function(x, key) {
+    genomesList <- getGenomesContentList(trackhub(x))
+    position <- which(sapply(genomesList, function(y) genome(x) %in% y))
+    genomesList[[position]][[key]]
+}
+
+Genome <- function(genome = "", trackDb = "", twoBitPath = "", groups = "",
+                   description = "", organism = "", defaultPos = "",
+                   orderKey = "", scientificName = "", htmlPath = "") {
+        c(as.list(environment()))
+}
+
+setGenomeContentList <- function(x, file) {
+    sapply(names(x), function(y) {
+        if(x[[y]] != "")
+            cat(y, " ", x[[y]], "\n", append = TRUE, sep = "", file = file)
+    })
+    cat("\n", append = TRUE, file = file)
+}
+
+setHubContent <- function(x, hubContent) {
+    cat("", file = hubFile(x))
+    sapply(names(hubContent), function(y) {
+        if (!is.null(hubContent[y]) && !is.na(hubContent[y]))
+            cat(y, " ", hubContent[y], "\n", append = TRUE, sep = "", file = hubFile(x))
+    })
+    cat("\n", append = TRUE, file = hubFile(x))
+}
+
+setMethod("uri", "TrackHub", function(x) {
+    x@uri
+})
+
+setMethod("hub", "TrackHub", function(x) {
+    hubContent <- getHubContent(hubFile(x))
+    hubContent[["hub"]]
+})
+
+setReplaceMethod("hub", "TrackHub", function(x, value) {
+    stopIfNotLocal(hubFile(x))
+    hubContent <- getHubContent(hubFile(x))
+    hubContent["hub"] <- value
+    setHubContent(x, hubContent)
+    x
+})
+
+setMethod("shortLabel", "TrackHub", function(x) {
+    hubContent <- getHubContent(hubFile(x))
+    hubContent[["shortLabel"]]
+})
+
+setReplaceMethod("shortLabel", "TrackHub", function(x, value) {
+    stopIfNotLocal(hubFile(x))
+    hubContent <- getHubContent(hubFile(x))
+    hubContent["shortLabel"] <- value
+    setHubContent(x, hubContent)
+    x
+})
+
+setMethod("longLabel", "TrackHub", function(x) {
+    hubContent <- getHubContent(hubFile(x))
+    hubContent[["longLabel"]]
+})
+
+setReplaceMethod("longLabel", "TrackHub", function(x, value) {
+    stopIfNotLocal(hubFile(x))
+    hubContent <- getHubContent(hubFile(x))
+    hubContent["longLabel"] <- value
+    setHubContent(x, hubContent)
```

Comment:
Rather than writing to file on every modification, please consider just modifying the object in memory, and then requiring an explicit save operation after all modifications. That better fits the R model, where side effects are explicit.

Reply:
okay understood, I will add an accessor method `save()` instead of implicitly saving the content.

Reply:
`save()` is already taken by the base package, but something like `writeTrackHub()` would make sense.